pkg/controller/attach_detach.go

```diff
@@ -599,128 +585,130 @@ func (r *ReconcileAttachDetach) recreateAzVolumeAttachment(ctx context.Context,
-				w.Logger().Infof("Found AzVolumeAttachment (%s) with older api version. Converting to apiVersion(%s)", azVolumeAttachmentName, azdiskv1beta2.APIVersion)
-
-				for k, v := range desiredAzVolumeAttachment.Labels {
-					azVolumeAttachment.Labels = azureutils.AddToMap(azVolumeAttachment.Labels, k, v)
-				}
+				return syncedVolumeAttachments, volumeHandles, volumeAttributes, err
+			}
+			syncedVolumeAttachments[volumeAttachment.Name] = true
+		}
+	}
+	return syncedVolumeAttachments, volumeHandles, volumeAttributes, err
+}
 
-				for k, v := range azVolumeAttachment.Annotations {
-					azVolumeAttachment.Status.Annotations = azureutils.AddToMap(azVolumeAttachment.Status.Annotations, k, v)
-				}
+// recreateAzVolumeAttachmentWithReplicaRole fetches attachment information from VM to properly recreate replica AzVolumeAttachments
+func (r *ReconcileAttachDetach) recreateAzVolumeAttachmentWithReplicaRole(
+	ctx context.Context,
+	syncedAzVolumeAttachments map[string]bool,
+	volumeHandles map[string]string,
+	volumeAttributes map[string]map[string]string) (map[string]bool, error) {
+	// In a normal case, all replica AzVolumeAttachments should already be detached and deleted on driver uninstall. So nothing needs to be recreated here.
+	// But in case of any failed detachment on previous driver uninstall, this function is needed.
+	var err error
+	w, _ := workflow.GetWorkflowFromContext(ctx)
+	nodes, err := r.kubeClient.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		w.Logger().Errorf(err, "failed to get list of nodes")
+		return syncedAzVolumeAttachments, err
+	}
 
-				// for now, we don't empty the meta annotatinos after migrating them to status annotation for safety.
-				// note that this will leave some remnant garbage entries in meta annotations
+	if syncedAzVolumeAttachments == nil {
+		syncedAzVolumeAttachments = map[string]bool{}
+	}
 
-				for k, v := range desiredAzVolumeAttachment.Annotations {
-					azVolumeAttachment.Annotations = azureutils.AddToMap(azVolumeAttachment.Status.Annotations, k, v)
-				}
+	for _, node := range nodes.Items {
+		nodeName := node.Name
+		// Get all attached disks from VM
+		disks, _, err := r.cloudDiskAttacher.GetNodeDataDisks(types.NodeName(nodeName), azcache.CacheReadTypeDefault)
+		if err != nil {
+			w.Logger().Errorf(err, "failed to get data disks for node (%s).", nodeName)
+			continue
+		}
 
-				azVolumeAttachment, err = r.azClient.DiskV1beta2().AzVolumeAttachments(r.config.ObjectNamespace).Update(ctx, azVolumeAttachment, metav1.UpdateOptions{})
+		for _, dataDisk := range disks {
+			azVolumeAttachmentName := azureutils.GetAzVolumeAttachmentName(*dataDisk.Name, nodeName)
+			if syncedAzVolumeAttachments[azVolumeAttachmentName] {
+				continue
+			}
+			if _, ok := r.azVolumeAttachmentToVaMap.Load(azVolumeAttachmentName); ok {
+				// If the AzVolumeAttachment has a VolumeAttachment mapping to it, it means it's a Primary role which should be recreated already. Skip it.
```

Comment:
Should we mark this volume as synced in this case?

Reply:
Done.